File: 2023/day1/day1.cs
#!/usr/bin/env dotnet-script
using System.Text.RegularExpressions;

string input = GetInput();

// input = @"
// 1abc2
// pqr3stu8vwx
// a1b2c3d4e5f
// treb7uchet";

// input = @"
// two1nine
// eightwothree
// abcone2threexyz
// xtwone3four
// 4nineeightseven2
// zoneight234
// 7pqrstsixteen
// ";

Console.WriteLine("part1: " + Day1(input));
Console.WriteLine("part2: " + Day2(input));

public static int Day1(string input)
{
  string[] lines = input.Trim().Split("\n");
  var sum = 0;
  foreach (var line in lines)
  {
    var matches = Regex.Matches(line, @"\d").OfType<Match>().Select(m => (m.Index, int.Parse(m.Value))).ToList();
    var (firstIndex, first) = matches.FirstOrDefault();
    var (lastIndex, last) = matches.LastOrDefault();
    var num = $"{first}{last}";
    sum += int.Parse(string.IsNullOrEmpty(num) ? "0" : num);
  }
  return sum;
}

public static int Day2(string input)
{
  string[] lines = input.Trim().Split("\n");
  var sum = 0;
  var numbers = new List<string> { string.Empty, "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };
  foreach (var line in lines)
  {
    var matches = Regex.Matches(line, @"\d").OfType<Match>().Select(m => (m.Index, int.Parse(m.Value))).ToList();
    var matchesWords = numbers.SelectMany((n, i) => line.AllIndexesOf(n).Select(index => (index, i))).ToList();
    matches.AddRange(matchesWords);
    matches = matches.OrderBy(x => x.Item1).ToList();
    var (firstIndex, first) = matches.FirstOrDefault();
    var (lastIndex, last) = matches.LastOrDefault();
    var num = $"{first}{last}";
    sum += int.Parse(string.IsNullOrEmpty(num) ? "0" : num);
  }
  return sum;
}

public static List<int> AllIndexesOf(this string str, string target)
{
  if (target == null) throw new ArgumentException("Target cannot be null", "target");
  if (target.Length == 0) return new();
  var indexes = new List<int>();
  for (int index = 0; ; index += target.Length)
  {
    index = str.IndexOf(target, index);
    if (index == -1) return indexes;
    indexes.Add(index);
  }
}

public static string GetInput()
{
  string command = System.Environment.CommandLine;
  string filename = command.Split(' ')[1];
  string puzzlePath = Path.Join(System.Environment.CurrentDirectory, filename);
  string inputPath = Path.Join(System.IO.Path.GetDirectoryName(puzzlePath), "input.txt");
  string input = File.ReadAllText(inputPath);
  return input;
}

File: 2023/day1/day1.py
from pathlib import Path
import re

def part1(input):
  lines = input.split('\n')
  sum = 0 
  for line in lines:
    matches = re.findall(r"\d", line)
    if len(matches) == 0:
      continue
    sum += int(str(matches[0] + matches[-1]))
  return sum

def part2(input):
  lines = input.split('\n')
  numbers = ['','one','two','three','four','five','six','seven','eight','nine'] 
  sum = 0 
  for line in lines:
    matchesInt = [[m.span()[0], m.group()] for m in re.finditer(r'\d', line)]
    matchWords = [[m.span()[0], str(numbers.index(m.group(1)))] for m in re.finditer(r'(?=(one|two|three|four|five|six|seven|eight|nine))', line)]
    matches = matchesInt + matchWords
    if len(matches) == 0:
      continue
    matches.sort(key=lambda pair: pair[0])
    # matches = list(map(lambda m: m[1], matches))
    matches = [m[1] for m in matches] 
    sum += int(str(matches[0] + matches[-1]))
  return sum

if __name__ == '__main__':
  inputPath = Path(__file__).parent.resolve() / 'input.txt'
  input = Path(inputPath).read_text().strip()

#   input = """
# two1nine
# eightwothree
# abcone2threexyz
# xtwone3four
# 4nineeightseven2
# zoneight234
# 7pqrstsixteen
#   """.strip()

  print(f'Part 1: {part1(input)}')
  print(f'Part 2: {part2(input)}')
